src/Microsoft.AspNetCore.Hosting/Internal/ApplicationLifetime.cs

```diff
@@ -44,13 +55,27 @@ public void StopApplication()
             // which will no-op since the first call already requested cancellation, get a chance to execute.
             lock (_stoppingSource)
             {
+                // Noop if this is already cancelled, user code can call this so 
+                // we should guard against multiple calls here
+                if (_stoppingSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
                 try
                 {
                     _stoppingSource.Cancel(throwOnFirstException: false);
+
+                    foreach (var handler in _handlers)
```

Comment:
How should we handle errors thrown here? Do we want 2 layers of try catch? If errors happen invoking cancellation token callbacks, should events get called?